habitat-lab/habitat/datasets/rearrange/navmesh_utils.py

```diff
@@ -10,6 +10,137 @@
 from habitat_sim.physics import VelocityControl
 
 
+def snap_point_is_occluded(
+    target: mn.Vector3,
+    snap_point: mn.Vector3,
+    height: float,
+    sim: habitat_sim.Simulator,
+    granularity: float = 0.2,
+    target_object_id: Optional[int] = None,
+) -> bool:
+    """
+    Uses raycasting to check whether a target is occluded given a navmesh snap point.
+
+    :property target: The 3D position whic should be unoccluded from the snap point.
+    :property snap_point: The navmesh snap point under consideration.
+    :property height: The height of the agent. Given navmesh snap point is grounded, the maximum height from which a visibility check should indicate non-occlusion. First check starts from this height.
+    :property sim: The Simulator instance.
+    :property granularity: The distance between raycast samples. Finer granularity is more accurate, but more expensive.
+    :property target_object_id: An optional object id which should be ignored in occlusion check.
+
+    NOTE: If agent's eye height is known and only that height should be considered, provide eye height and granulatiry > height for fastest check.
+
+    :return: whether or not the target is considered occluded from the snap_point.
+    """
+    # start from the top, assuming the agent's eyes are not at the bottom.
+    cur_height = height
+    while cur_height > 0:
+        ray = habitat_sim.geo.Ray()
+        ray.origin = snap_point + mn.Vector3(0, cur_height, 0)
```

Comment:
I think this boils down to where the camera's location is. So Spot, we will need to offset the `base_pos` to the front and then set the height to be ee's height?

Reply:
Yeah, exactly. Height should be the maximum height where the robot hand/camera can have valid visibility to the object. Max EE camera height may be too much (e.g. the robot can't grasp an object it can only see with it's arm stretched all the way up).